src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java

```diff
@@ -1,46 +1,100 @@
-        products.add(new Product(13L, "simsxg75", "Siemens SXG75", new BigDecimal(150), usd, 40, "https://raw.githubusercontent.com/andrewosipenko/phoneshop-ext-images/master/manufacturer/Siemens/Siemens%20SXG75.jpg"));
     }
 
     @Override
     public Product getProduct(Long id) {
         return products.stream()
                 .filter(s->s.getId().equals(id) && s.getPrice()!=null && s.getStock()>0)
                 .findAny()
-                .orElse(null);
+                .orElseThrow(()-> new ProductNotFoundException(id));
     }
 
     @Override
-    public List<Product> findProducts() {
-        return products.stream()
+    public List<Product> findProducts(String query, SortField sortField,SortOrder sortOrder) {
+
+        List<Product> result =products.stream()
                 .filter(s-> s.getPrice()!=null && s.getStock()>0)
                 .collect(Collectors.toList());
+
+        if (query!=null) {
+            Map<Product,Integer> productsSearch=new TreeMap<>();
+
+            String[] partQuery = query.split(" ");
+
+            result.stream()
+                    .filter(s ->
+                    {
+                        int i = 0;
+                        for (String part : partQuery) {
+                            if (s.getDescription().contains(part))
+                                i++;
+                        }
+                        if (i == 0) {
+                            return false;
+                        } else {
+                            productsSearch.put(s, i);
+                            return true;
+                        }
+                    })
+                    .collect(Collectors.toList());
+
+            productsSearch.entrySet().stream().sorted(Map.Entry.comparingByValue());
+            result.clear();
+            result.addAll(productsSearch.keySet());
+        }
+        if (sortField==SortField.PRICE) {
+            if (sortOrder == SortOrder.ASC) {
+                return result.stream()
+                        .sorted(Comparator.comparing(Product::getPrice))
+                        .collect(Collectors.toList());
+            }
+
+            else {
+                return result.stream()
+                        .sorted(Comparator.comparing(Product::getPrice, Comparator.reverseOrder()))
+                        .collect(Collectors.toList());
+            }
+        }
+        if (sortField==SortField.DESCRIPTION){
+            if (sortOrder==SortOrder.ASC) {
+                return result.stream()
+                        .sorted(Comparator.comparing(Product::getDescription))
+                        .collect(Collectors.toList());
+            }
+            else {
+                return result.stream()
+                        .sorted(Comparator.comparing(Product::getDescription, Comparator.reverseOrder()))
+                        .collect(Collectors.toList());
+            }
+        }
+        return result;
+
```

Comment:
1. The method is too long and it's difficult to read it. Can we split it into a few smaller methods: the first to search for query, the second to sort the list.

2. Instead of using 4 similar code chunks like:

```
result.stream()
                        .sorted(Comparator.comparing(...), ...)
                        .collect(Collectors.toList());
```

can we put this in a separate method? The method can accept the field name and sort direction and return the comparator.

Something like:

```
private Comparator<Product> getComparator(String order, String sort) {
    Comparator<Product> comparator = Comparator.comparing(...by passed field...);
    ...return comparator or comparator.reversed()...
}
```

and then utilize this method's return value:

```
product..stream()
                .sorted(getComparator(order, sort))
                .collect(Collectors.toList());
```